spec/client/client/BootstrappedDataServiceSpec.js

```diff
@@ -16,90 +15,120 @@ describe("BootstrappedDataService", function() {
-
-            it("mocks out the next ajax call which will happen when sync completes", function(done) {
                 bootstrappedDataService.checkAlreadyHasData(method, model, options);
-
-                ajaxRequestModelData().then(function(data) {
-                    expect(data).toEqual({
-                        "some": "data"
-                    });
-
-                    done();
-                });
+                makeAjaxRequest = ajaxRequestModelData;
             });
 
-            describe("when data is requested twice", function() {
+            itMocksAjaxCallsFor("/url");
+        });
 
-                var fetchingTwice;
+        describe("when there is bootstrapped data and request has query params", function() {
+            beforeEach(function() {
+                options = {
+                    data: {
+                        query: "param"
+                    }
+                };
 
-                it("does NOT mock the second request", function(done) {
-                    bootstrappedDataService.checkAlreadyHasData(method, model, options);
+                bootstrappedDataService = new BootstrappedDataService(bootstrappedData);
+                bootstrappedDataService.checkAlreadyHasData(method, model, options);
+                makeAjaxRequest = ajaxRequestModelDataWithQueryParams;
+            });
 
-                    fetchingTwice = ajaxRequestModelData()
-                        .then(function() {
-                            bootstrappedDataService.checkAlreadyHasData(method, model, options);
+            itMocksAjaxCallsFor('/url{"query":"param"}');
+        });
 
-                            return ajaxRequestModelData();
-                        });
+        describe("when there is no bootstrapped data and request does NOT have query params", function() {
+            beforeEach(function() {
+                options = {};
+                bootstrappedDataService = new BootstrappedDataService();
+                bootstrappedDataService.checkAlreadyHasData(method, model, options);
+                makeAjaxRequest = ajaxRequestModelData;
+            });
 
-                    // using always because this is a jquery promise AND it fails on the second
-                    //  call since an ajax call to /url in the tests won't succeed. this
-                    //  expectation only cares that the mock data is NOT returned again so that's ok.
-                    //  wawjr3d 10/17/2014
-                    fetchingTwice.always(function(data) {
-                        expect(data).not.toEqual({
-                            "some": "data"
-                        });
+            itDoesNotMockAjaxCallsFor("/url");
+        });
 
-                        done();
-                    });
-                });
+        describe("when there is no bootstrapped data and request has query params", function() {
+            beforeEach(function() {
+                options = {
+                    data: {
```

Comment:
can you add tests for when data is Array or String. jQuery docs say those are allowed too